pkg/admission/annotations.go

```diff
@@ -0,0 +1,44 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	LastDownscaleAnnotationKey = "last-prepared-for-downscale"
+)
+
+func addPreparedForDownscaleAnnotationToPod(ctx context.Context, api kubernetes.Interface, namespace, stsName string, podNr int) error {
+	client := api.CoreV1().Pods(namespace)
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{
+		"name":                               stsName,
+		"statefulset.kubernetes.io/pod-name": fmt.Sprintf("%v-%v", stsName, podNr),
+	}}
+	pods, err := client.List(ctx,
+		v1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		return err
+	}
+	if len(pods.Items) != 1 {
+		return fmt.Errorf("multiple or no pods found for statefulset %v and index %v", stsName, podNr)
+	}
+
+	pod := pods.Items[0]
+
+	annotations := pod.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	annotations[LastDownscaleAnnotationKey] = time.Now().UTC().String()
```

Comment:
For my own clarification, just the presence of this annotation is what matters to a running application, right? The fact that the annotation exists means this pod should expect to be permanently shut down. The date here is for the benefit of us (the operators). If this is the case, could you note that in comments or the README somewhere.

Reply:
A comment was added.